Mathlib/Topology/SubsetProperties.lean

```diff
@@ -1307,15 +1309,151 @@ theorem IsClosed.exists_minimal_nonempty_closed_subset [CompactSpace α] {S : Se
   rw [← this, compl_compl]
 #align is_closed.exists_minimal_nonempty_closed_subset IsClosed.exists_minimal_nonempty_closed_subset
 
+/-- A subset `s ⊆ α` is called **σ-compact** if it is the union of countably many compact sets. -/
+def IsSigmaCompact (s : Set α) : Prop :=
+  ∃ K : ℕ → Set α, (∀ n, IsCompact (K n)) ∧ ⋃ n, K n = s
+
+/-- The empty set is σ-compact. -/
+@[simp]
+lemma isSigmaCompact_empty : IsSigmaCompact (∅ : Set α) := by
+  use fun _ ↦ ∅
+  simp only [isCompact_empty, forall_const, iUnion_empty, and_self]
+
+/-- Compact sets are σ-compact. -/
+lemma IsCompact.isSigmaCompact {s : Set α} (hs : IsCompact s) : IsSigmaCompact s :=
+  ⟨fun _ => s, fun _ => hs, iUnion_const _⟩
+
+/-- Countable unions of compact sets are σ-compact. -/
+lemma isSigmaCompact_sUnion_of_isCompact {S : Set (Set α)} (hc : Set.Countable S)
+    (hcomp : ∀ (s : Set α), s ∈ S → IsCompact s) : IsSigmaCompact (⋃₀ S) := by
+  by_cases S = ∅
+  · simp only [h, sUnion_empty, isSigmaCompact_empty]
+  · -- If S is non-empty, choose a surjection f : ℕ → S, this yields a map ℕ → Set X.
+    obtain ⟨f, hf⟩ := (Set.countable_iff_exists_surjective (nmem_singleton_empty.mp h)).mp hc
+    refine ⟨fun n ↦ f n, fun n ↦ hcomp (f n) (Subtype.mem (f n)), ?_⟩
+    rw [Function.Surjective.iUnion_comp hf]
+    exact sUnion_eq_iUnion.symm
+
+/-- Countable unions of σ-compact sets are σ-compact.
+  Formulated here for an indexed union of sets in a given set. -/
+lemma isSigmaCompact_iUnion {ι : Type*} [Countable ι] (s : ι → Set α)
+    (hcomp : ∀ i, IsSigmaCompact (s i)) : IsSigmaCompact (⋃ i, s i) := by
+  -- Choose a decomposition s = ⋃ s_i for each s ∈ S.
+  choose K hcomp hcov using fun i ↦ hcomp i
+  -- Then, we have a countable union of countable unions of compact sets, i.e. countably many.
+  have : ⋃ i, s i = ⋃₀ range (K.uncurry) := by calc ⋃ i, s i
+      _ = ⋃ i, ⋃ n, (K i n) := by simp_rw [hcov]
+      _ = ⋃ (i) (n : ℕ), (K.uncurry ⟨i, n⟩) := by rw [Function.uncurry_def]
+      _ = ⋃₀ range (K.uncurry) := by rw [← iUnion_prod', sUnion_range]
+  rw [this]
+  refine isSigmaCompact_sUnion_of_isCompact (countable_range (K.uncurry)) fun s hs ↦ ?_
+  obtain ⟨⟨ys, yn⟩, hy⟩ := mem_range.mp hs
+  exact hy ▸ hcomp ys yn
```

Comment:
This still looks a bit too complicated. If you add the following (either replacing `sUnion` version you already have, or using the `iUnion` version to prove the `sUnion` one):
```lean
lemma isSigmaCompact_iUnion_of_isCompact {ι : Type*} [hι : Countable ι] (s : ι → Set α)
    (hcomp : ∀ i, IsCompact (s i)) : IsSigmaCompact (⋃ i, s i) := by
  rcases isEmpty_or_nonempty ι
  · simp only [iUnion_of_empty, isSigmaCompact_empty]
  · -- If ι is non-empty, choose a surjection f : ℕ → ι, this yields a map ℕ → Set X.
    obtain ⟨f, hf⟩ := countable_iff_exists_surjective.mp hι
    exact ⟨s ∘ f, fun n ↦ hcomp (f n), Function.Surjective.iUnion_comp hf _⟩
```
You can do:
```suggestion
      _ = ⋃ x, K.uncurry x := by rw [← iUnion_prod']
  rw [this]
  exact isSigmaCompact_iUnion_of_isCompact K.uncurry fun x ↦ (hcomp x.1 x.2)
```
But maybe I'm nit-picking too much...

Reply:
If you prefer this, I don't mind. Added.